### ISSUE TITLE
Great difference results of generalized linear model between xgboost and glmnet

### ISSUE THREAD

**Issue by bifeng** [dmlc/xgboost#2098]
The xgboost for glm code below:
require(xgboost)
data(agaricus.train, package='xgboost')
data(agaricus.test, package='xgboost')
dtrain <- xgb.DMatrix(agaricus.train$data, label = agaricus.train$label)
dtest <- xgb.DMatrix(agaricus.test$data, label = agaricus.test$label)
param <- list(objective = "binary:logistic", booster = "gblinear",
              nthread = 2, alpha = 0.0001, lambda = 1)
num_round <- 1
bst <- xgb.train(param, dtrain, num_round)
importance_matrix <- xgb.importance(colnames(agaricus.train$data), model = bst)
(gg <- xgb.ggplot.importance(importance_matrix, measure = "Weight", rel_to_first = TRUE))
gg + ggplot2::ylab("Weight")
ypred <- predict(bst, dtest)
labels <- getinfo(dtest, 'label')
cat('error of preds=', mean(as.numeric(ypred>0.5)!=labels),'\n')
# error of preds= 0.01675978 
head(importance_matrix)
![image](https://cloud.githubusercontent.com/assets/8215563/23820521/4e87613e-0655-11e7-9c5d-4eed60ac402c.png)

The glmnet for glm code below:
require(glmnet)
require(varhandle)
trainx <- as.matrix(agaricus.train$data)
trainy <- factor(agaricus.train$label, labels = c("A","B"))
testx <- as.matrix(agaricus.test$data)
testy <- factor(agaricus.test$label, labels = c("A","B"))
testy2 <- agaricus.test$label

glm_result <- glmnet(trainx, trainy, family = "binomial", alpha = 0.0001, lambda = 1)
glm_result
coef_importance <- coef(glm_result, s=1)
pre_result <- predict(glm_result, newx = testx, s = 1,  type = "class")
pre_result2 <- as.numeric(unfactor(factor(pre_result, labels = c('0','1'))))
cat('error of preds=', mean(pre_result2!=testy2),'\n')
# error of preds= 0.02979516 
head(coef_importance)
![image](https://cloud.githubusercontent.com/assets/8215563/23820533/8d52a5e0-0655-11e7-8dbe-313865ddb92e.png)

# There is so much difference between each others. The glmnet's prediction error higher than xgboost  even the same paramters! 
# how to explain the difference between coef (glmnet) and weight (xgboost) ? 

> sessionInfo()
R version 3.3.1 (2016-06-21)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

locale:
[1] LC_COLLATE=Chinese (Simplified)_China.936 
[2] LC_CTYPE=Chinese (Simplified)_China.936   
[3] LC_MONETARY=Chinese (Simplified)_China.936
[4] LC_NUMERIC=C                              
[5] LC_TIME=Chinese (Simplified)_China.936    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods  
[7] base     

other attached packages:
[1] varhandle_2.0.1 glmnet_2.0-5    foreach_1.4.3   Matrix_1.2-6   
[5] xgboost_0.6-4  

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.7           lattice_0.20-33       codetools_0.2-14     
 [4] digest_0.6.10         plyr_1.8.4            chron_2.3-47         
 [7] grid_3.3.1            gtable_0.2.0          magrittr_1.5         
[10] scales_0.4.0          ggplot2_2.1.0         stringi_1.1.1        
[13] data.table_1.9.6      labeling_0.3          iterators_1.0.8      
[16] tools_3.3.1           Ckmeans.1d.dp_3.4.6-1 munsell_0.4.3        
[19] rsconnect_0.4.3       colorspace_1.2-6

**Comment by khotilov** [comment]
They are two rather different models in terms of how they were built. And their parameters are not even comparable. In other words, you threw some rather random sets of parameters (that also were far from optimal) at two fairly different algorithms - no wonder the results were not the same.

**Comment by bifeng** [comment]
@khotilov Thank you for you quicklly response! Can you explain the weight calculation method in the xgboost for gblinear booster? It seems rarely comment or papers about it.

**Comment by bifeng** [comment]
In this site: https://www.kaggle.com/general/25924#post148061, there are following words about the gblinear:
Put it simply it is really a "regularized linear model" using delta with elastic net regularization (L1 + L2 + L2 bias) and parallel coordinate descent optimization.
So, I think the model difference between xgboost's gblinear and glmnet is not so much, if we can set the same parameters.

**Comment by khotilov** [comment]
I should have phrased my earlier response more clearly. It should have been: "The two models produced in your example are the results of two different processes. In large part, that difference comes from your parameter choices. There's also a fair difference in the algorithms... But the main reasons seem to be that you don't quite understand what each of those parameters means (the same names != the same meaning) and that you are comparing largely suboptimal models."

Anyway, glmnet uses [cyclical gradient descent](https://web.stanford.edu/~hastie/glmnet/glmnet_alpha.html). When running in a single thread mode, gblinear also does a similar "cycle" of gradient updates at each iteration. However gradient boosting iterations work their way in a fairly different manner than the iterations in glmnet. Moreover, when running multithreaded, there's some hogwild (non-thread-safe) parallelization happening. 

As for the parameters, please carefully read the documentation of both packages. Also keep in mind that the loss in xgboost is not normalized to #samples, which means that the penalty parameters would differ by a factor of N. Plus, you would have to run gblinear through quite a lot of iterations to get the coefficients comparable between the two models.
